backend/models/notification.js

```diff
@@ -0,0 +1,255 @@
+		return notifications.map(notification => {
+			const teamSpace = (modelsData[notification.teamSpace] || {});
+			const modelName = (teamSpace[notification.modelId] || {}).name;
+			return Object.assign(notification, {modelName});
+		});
+	});
+};
+
+const getNotification = (username, type, criteria) =>
+	db.getCollection(NOTIFICATIONS_DB, username)
+		.then((collection) => collection.find(Object.assign({type},  criteria)).toArray());
+
+const convertIfNecessary = function(_id) {
+	return _.isString(_id) ? utils.stringToUUID(_id) : _id;
+};
+
+module.exports = {
+	types,
+
+	/**
+	 * Creates a notification in the database
+	 *
+	 * @param {string} username The username of the account thats's gonna receive the notification
+	 * @param {string} type	The type of notification: should be one of the notifications that is in the types constants
+	 * @param {Object} data The particular data for notification. should be relevant data for the particular type of notification.
+	 * @returns {Promise} Returns a promise with the recently created notification
+	 */
+	insertNotification: function(username, type, data) {
+		return db.getCollection(NOTIFICATIONS_DB, username).then((collection) =>
+			collection.insertOne(generateNotification(type, data))
+		).then((o) => utils.changeObjectIdToString(o.ops[0]));
+	},
+
+	updateNotification: function(username, _id, data) {
+		_id = convertIfNecessary(_id);
+		return db.getCollection(NOTIFICATIONS_DB, username).then((collection) =>
+			collection.update({_id}, { $set: data })
+		);
+	},
+
+	upsertNotification: function(username, data, type, criteria) {
+		return getNotification(username, type, criteria).then(notifications => {
+			if (notifications.length === 0) {
+				return this.insertNotification(username, type, Object.assign(criteria, data));
+			} else {
+				const n = notifications[0];
+				const timestamp = (new Date()).getTime();
+				const mergedData = Object.assign(_.mergeWith(n, data, unionArrayMerger), {read:false,timestamp});
+				return this.updateNotification(username, n._id, mergedData).then(() => {
+					const notification =  Object.assign(n, mergedData);
+					return utils.changeObjectIdToString(notification);
+				});
+			}
+		});
+	},
+
+	upsertIssueAssignedtNotification: function(username, teamSpace, modelId, issueId) {
+		const criteria = {teamSpace,  modelId};
+		const data = {issuesId: [issueId] };
+		return this.upsertNotification(username,data,types.ISSUE_ASSIGNED,criteria);
+	},
+
+	removeIssueAssignedNotification:function(username, teamSpace, modelId, issueId) {
+		const criteria = {teamSpace,  modelId, issuesId:{$in: [issueId]}};
+
+		return getNotification(username, types.ISSUE_ASSIGNED, criteria).then(notifications => {
+			if (notifications.length === 0) {
+				return null;
+			} else {
+				const n = notifications[0];
+				const index = n.issuesId.findIndex(i => i === issueId);
```

Comment:
Is there a scenario where this returns -1?

like, if I got the notification, removed it, then someone unassigned me?

Reply:
🤔 I dont think so, cause the search a little bit more above has the search criteria of searching for the notifications that have that specific issueId  in the issuesId array.